File: c++/003.Longest Substring Without Repeating Characters.cpp
class Solution{
public:
	int lengthOfLongestSubstring(string s){
		//使用 two point 和 hashTable 方法
		int i = 0, j = 0, len = s.length();
		int maxlen = 0;		//最大长度
		int hashTable[128] = {0};		//标记字母a~z, A~Z是否出现过
		while(j < len){
			if(hashTable[s[j]] == 0){		//字母还未出现
				hashTable[s[j]] = 1;
			}else{			// 字母出现过
				maxlen = max(maxlen, j - i);
				while(s[i] != s[j]){		
					hashTable[s[i]] = 0;
					i++;
				}
				i++;
			}
			j++;
		}
		return max(maxlen, j-i);
	}
};

File: Java/003.Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
    	// maxlen 用来记录子串的最大长度
    	// pos 用来记录最后一个重复字符的位置, 初始为 0
        int maxlen = 0, pos = 0;
        // index 数组表示当前字符的索引
        int[] index = new int[128];

        for(int i = 0; i < s.length(); i++){
        	pos = Math.max(index[s.charAt(i)], pos);
        	maxlen = Math.max(maxlen, i - pos + 1);
        	index[s.charAt(i)] = i + 1;
        }
        return maxlen;
    }
}